File: graph/14.cpp
#include <bits/stdc++.h>

using namespace std;

using Matrix = vector<vector<int>>;
using Grid = vector<vector<char>>;

bool isValid(const Grid &G, const Matrix &visited, int row, int col)
{

    return row >= 0 && row < G.size() && col >= 0 && col < G[0].size() && visited[row][col] == 0 && G[row][col] == '1';
}

void DFS(Grid &G, Matrix &visited, pair<int, int> coor)
{
    vector<int> posRow = {-1, 0, 0, 1};
    vector<int> posCol = {0, -1, 1, 0};

    visited[coor.first][coor.second] = 1;

    for (int i = 0; i < posRow.size(); ++i)
    {
        int newPosRow = coor.first + posRow[i];
        int newPosCol = coor.second + posCol[i];

        if (isValid(G, visited, newPosRow, newPosCol))
            DFS(G, visited, {newPosRow, newPosCol});
    }
}

int numIslands(vector<vector<char>> &grid)
{
    Matrix visited(grid.size(), vector<int>(grid[0].size(), 0));
    int cont = 0;

    for (int i = 0; i < grid.size(); ++i)
    {
        for (int j = 0; j < grid[i].size(); ++j)
        {
            if (visited[i][j] == 0 && grid[i][j] == '1')
            {
                cont++;
                DFS(grid, visited, {i, j});
            }
        }
    }

    return cont;
}

int main()
{
    Grid G(4);
    G[0] = {'1', '1', '0', '0', '0'};
    G[1] = {'1', '1', '0', '0', '0'};
    G[2] = {'0', '0', '1', '0', '0'};
    G[3] = {'0', '0', '0', '1', '1'};

    cout << numIslands(G);

    return 0;
}

File: graph/14.py
from typing import List


def is_valid(grid, visited, row, col):
    return (
        row >= 0
        and row < len(grid)
        and col >= 0
        and col < len(grid[0])
        and visited[row][col] == 0
        and grid[row][col] == "1"
    )


def DFS(grid, visited, row, col):
    visited[row][col] = 1

    posRow = [-1, 0, 0, 1]
    posCol = [0, -1, 1, 0]

    for index in range(len(posRow)):
        newPosRow = row + posRow[index]
        newPosCol = col + posCol[index]

        if is_valid(grid, visited, newPosRow, newPosCol):
            DFS(grid, visited, newPosRow, newPosCol)


def numIslands(grid: List[List[str]]) -> int:
    visited = [[0] * len(row) for row in grid]
    cont = 0

    for row in range(len(grid)):
        for col in range(len(grid[row])):
            if visited[row][col] == 0 and grid[row][col] == "1":
                cont += 1
                DFS(grid, visited, row, col)

    return cont


grid = [
    ["1", "1", "0", "0", "0"],
    ["1", "1", "0", "0", "0"],
    ["0", "0", "1", "0", "0"],
    ["0", "0", "0", "1", "1"],
]

print(numIslands(grid))
